### PR TITLE
feat: add pref change subscriptions

### PR DESCRIPTION
**Pull request by aminomancer**

### CODE REVIEW

File: src/content/components/MyBugs/MyBugs.js [hunk 1]
```diff
@@ -31,9 +31,12 @@ export class MyBugs extends React.PureComponent {
       showSettings: false,
       emailWasChanged: false,
     };
+    this.onPrefChange = this.onPrefChange.bind(this);
     this.onEmailChange = this.onEmailChange.bind(this);
     this.onEmailSubmit = this.onEmailSubmit.bind(this);
     this.toggleSettings = this.toggleSettings.bind(this);
+
+    prefs.on("bugzilla_email", this.onPrefChange);
```
Comment (aminomancer) [inline src/content/components/MyBugs/MyBugs.js:39]: maybe this should be done in componentDidMount and removed in componentDidUnmount, though I didn't notice any errors doing it this way. with the priority guide I think you can do it the same way since I *think* it never gets unmounted

File: src/content/lib/prefs.js [hunk 2]
```diff
@@ -41,16 +41,83 @@ class Prefs {
...
   }
 
   reset(name) {
-    return this._store.delete(name);
+    const oldValue = this.get(name);
+    const newValue = this._defaults[name];
+    if (oldValue !== newValue) {
+      this._store.delete(name);
+      this._emit(name, oldValue, newValue);
+    }
```
Comment (rjfc) [inline src/content/lib/prefs.js:57]: What do you think about replacing this with `set(name, this._defaults[name]);`?

File: src/content/lib/prefs.js [hunk 3]
```diff
@@ -41,16 +41,83 @@ class Prefs {
...
+        this._apply(listener, { name, oldValue, newValue });
+      }
+    }
+  }
+
+  _apply(listener, data) {
+    if (typeof listener === "function") {
+      listener(data);
+    } else if (listener?.onPrefChange) {
+      listener.onPrefChange(data);
```
[the remark this answers is not in the document]
Comment (aminomancer) [reply src/content/lib/prefs.js:114]: It's just conventional for event listener APIs to support either callback functions or objects with a standard callback property (in this case, `onPrefChange`). See e.g. [EventListener::handleEvent](https://searchfox.org/mozilla-central/rev/11dbac7f64f509b78037465cbb4427ed71f8b565/dom/webidl/EventListener.webidl#15) (also see [addEventListener](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#the_event_listener_callback)) or [nsIObserver::observe](https://searchfox.org/mozilla-central/source/xpcom/ds/nsIObserver.idl).

File: src/content/lib/prefs.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (aminomancer) [reply src/content/lib/prefs.js:57]: That would permanently store the default value as the user value. Which isn't how default values are supposed to work. Consider for example what happens if we want to change the default value for a pref. But the previous default value has already been saved as a user value. So when something requests the value, it won't get the new default value, it'll get the user value (IOW, the old default value).

It doesn't really matter since I think we don't even use this method, but still not ideal.

Actually, this code as I have it is not quite right, either. I seem to have absentmindedly copied from the set method, but it really should delete the value no matter what. Since as it is right now, it will only delete the value if it differs from the default value. But that means it fails to delete a user value that equals the default value. Which prevents cleaning up unnecessary user values.

Basically we only want to emit an event if this resulted in the "effective value" changing, i.e. if deleting the value would result in `get()` returning a different value. If the user value was already equal to the default value, then resetting won't change the value, even though it _should_ delete the value from storage. So we want to delete from storage but only emit an event if the deleted value was non-default.

**Review by rjfc — APPROVED** [review]
